### DISCUSSION TITLE
Sync policy hooks and coming from other framework

### DISCUSSION THREAD

**Discussion by Xstoudi** [adonisjs/core#4378] [Help] [answered]
At work I work with Spring and this use case is rather trivial as accessing to `roles` will trigger it's retrieval automagically.

But I struggle a bit to do it with Adonis. `roles` (many to many relationship) is obviously not loaded here.
```
export default class UserPolicy extends BasePolicy {
  before(user: User | null) {
    if (user !== null && user.roles.some((role) => role.code === 'admin')) {
      return true
    }
  }

  view(_: User): AuthorizerResponse {
    return false
  }

  create(_: User): AuthorizerResponse {
    return false
  }

  edit(_: User): AuthorizerResponse {
    return false
  }

  delete(_: User): AuthorizerResponse {
    return false
  }
}
```

I could obviously add a hook on the `User` model to eager-load `roles`, but then they will be retrieved even when not needed. Is there a way to load `roles` in a fine-tuned way?

**Comment by tomgobich** [comment, marked as answer]
Hey there! Unfortunately, there isn't a great way within JavaScript to make the automatic retrieval of an asynchronous property work as it can in Spring.

Instead, you'll need to preload _(eager)_ or load _(lazy)_ the relationship, then you can access its properties. 
https://lucid.adonisjs.com/docs/relationships#lazy-load-relationships

For example:
```ts
export default class UserPolicy extends BasePolicy {
  async before(user: User | null) {
    if (user === null) return

    await user.load('roles')

    if (user.roles.some((role) => role.code === 'admin')) {
      return true
    }
  }

  // ...
}
```

Edit: Adding `async` to example `before` method

**Reply by Xstoudi** [reply]
I think it could be a nice addition to the doc to mention the fact that `before` an be async then. :D

Thanks!

**Reply by thetutlage** [reply]
Updated docs https://github.com/adonisjs/v6-docs/commit/6f44689abaacc1f550e45d2d8a56f320a245d76f
